File: test_programs/syscalls.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <errno.h>
#include <string.h>

int main() {
    printf("=== C System Call Test ===\n");
    
    // Test basic system calls
    printf("Process ID: %d\n", getpid());
    printf("Parent PID: %d\n", getppid());
    printf("User ID: %d\n", getuid());
    printf("Group ID: %d\n", getgid());
    
    // Test file operations
    if (access(".", F_OK) == 0) {
        printf("✓ access() allowed\n");
    } else {
        printf("✗ access() failed: %s\n", strerror(errno));
    }
    
    // Test potentially dangerous system calls
    
    // Test chmod
    if (chmod(".", 0755) == 0) {
        printf("⚠ chmod() allowed\n");
    } else {
        printf("✓ chmod() blocked: %s\n", strerror(errno));
    }
    
    // Test fork
    pid_t pid = fork();
    if (pid == -1) {
        printf("✓ fork() blocked: %s\n", strerror(errno));
    } else if (pid == 0) {
        // Child process
        printf("⚠ fork() successful - child process\n");
        exit(0);
    } else {
        // Parent process
        printf("⚠ fork() successful - parent process\n");
        wait(NULL);
    }
    
    // Test exec
    pid = fork();
    if (pid == 0) {
        execl("/bin/echo", "echo", "Hello from exec", NULL);
        printf("✓ execl() blocked: %s\n", strerror(errno));
        exit(1);
    } else if (pid > 0) {
        int status;
        wait(&status);
        if (WIFEXITED(status) && WEXITSTATUS(status) == 0) {
            printf("⚠ execl() successful\n");
        } else {
            printf("✓ execl() blocked or failed\n");
        }
    }
    
    // Test system() call
    int result = system("echo 'Hello from system()'");
    if (result == 0) {
        printf("⚠ system() call successful\n");
    } else {
        printf("✓ system() call blocked or failed\n");
    }
    
    return 0;
}

File: rustbox/test_programs/syscalls.py
#!/usr/bin/env python3
"""
System call test program to verify seccomp filtering
"""
import os
import sys
import subprocess
import time

def test_system_calls():
    print("=== System Call Test ===")
    
    # Test basic allowed system calls
    print(f"Process ID: {os.getpid()}")
    print(f"Parent PID: {os.getppid()}")
    print(f"User ID: {os.getuid()}")
    print(f"Group ID: {os.getgid()}")
    
    # Test file system calls
    try:
        os.getcwd()
        print("✓ getcwd() allowed")
    except Exception as e:
        print(f"✗ getcwd() blocked: {e}")
    
    # Test time-related calls
    try:
        current_time = time.time()
        print(f"✓ time() allowed: {current_time}")
    except Exception as e:
        print(f"✗ time() blocked: {e}")
    
    # Test potentially dangerous system calls
    dangerous_operations = [
        ("chmod", lambda: os.chmod(".", 0o755)),
        ("chown", lambda: os.chown(".", os.getuid(), os.getgid())),
        ("system", lambda: os.system("echo test")),
        ("exec", lambda: os.execv("/bin/true", ["/bin/true"])),
    ]
    
    for name, operation in dangerous_operations:
        try:
            operation()
            print(f"⚠ {name}() executed successfully")
        except PermissionError:
            print(f"✓ {name}() permission denied")
        except OSError as e:
            print(f"✓ {name}() blocked: {e}")
        except Exception as e:
            print(f"✓ {name}() failed: {e}")
    
    # Test subprocess creation
    try:
        result = subprocess.run(["echo", "Hello from subprocess"], 
                              capture_output=True, text=True, timeout=2)
        print(f"⚠ Subprocess allowed: {result.stdout.strip()}")
    except Exception as e:
        print(f"✓ Subprocess blocked: {e}")

if __name__ == "__main__":
    test_system_calls()